mining/src/mempool/model/frontier.rs

```diff
@@ -0,0 +1,391 @@
+            let item = {
+                let mut item = self.search_tree.search(query);
+                while !cache.insert(item.tx.id()) {
+                    _collisions += 1;
+                    if top == item {
+                        // Narrow the search to reduce further sampling collisions
+                        match down_iter.next() {
+                            Some(next) => top = next,
+                            None => break 'outer,
+                        }
+                        let remaining_weight = self.search_tree.prefix_weight(top);
+                        distr = Uniform::new(0f64, remaining_weight);
+                    }
+                    let query = distr.sample(rng);
+                    item = self.search_tree.search(query);
+                }
+                item
+            };
+            sequence.push(item.tx.clone(), item.mass);
+            total_selected_mass += item.mass; // Max standard mass + Mempool capacity bound imply this will not overflow
+        }
+        trace!("[mempool frontier sample inplace] collisions: {_collisions}, cache: {}", cache.len());
+        sequence
+    }
+
+    /// Dynamically builds a transaction selector based on the specific state of the ready transactions frontier.
+    ///
+    /// The logic is divided into three cases:
+    ///     1. The frontier is small and can fit entirely into a block: perform no sampling and return
+    ///        a TakeAllSelector
+    ///     2. The frontier has at least ~4x the capacity of a block: expected collision rate is low, perform
+    ///        in-place k*log(n) sampling and return a SequenceSelector
+    ///     3. The frontier has 1-4x capacity of a block. In this case we expect a high collision rate while
+    ///        the number of overall transactions is still low, so we take all of the transactions and use the
+    ///        rebalancing weighted selector (performing the actual sampling out of the mempool lock)
+    ///
+    /// The above thresholds were selected based on benchmarks. Overall, this dynamic selection provides
+    /// full transaction selection in less than 150 µs even if the frontier has 1M entries (!!). See mining/benches
+    /// for more details.  
+    pub fn build_selector(&self, policy: &Policy) -> Box<dyn TemplateTransactionSelector> {
+        if self.total_mass <= policy.max_block_mass {
+            Box::new(TakeAllSelector::new(self.search_tree.ascending_iter().map(|k| k.tx.clone()).collect()))
+        } else if self.total_mass > policy.max_block_mass * COLLISION_FACTOR {
+            let mut rng = rand::thread_rng();
+            Box::new(SequenceSelector::new(self.sample_inplace(&mut rng, policy), policy.clone()))
+        } else {
+            Box::new(RebalancingWeightedTransactionSelector::new(
+                policy.clone(),
+                self.search_tree.ascending_iter().cloned().map(CandidateTransaction::from_key).collect(),
+            ))
+        }
```

Comment:
The ordering here threw me off a bit, and I was a bit confused why the last one had an `ascending_iter`. Reading more into it, the second if-statement with the `SequenceSelector` is where the real magic of this PR happens (and not at the 3rd one).

Reply:
Right, the 2nd case is the interesting one. The `ascending_iter()` used in the 1rd and 3rd is just the default `iter()` method of the tree. The rebalancing selector is not sensitive to the order of input txs